x/mongo/driver/topology/pool.go

```diff
@@ -524,69 +405,104 @@ func (p *pool) getGenerationForNewConnection(serviceID *primitive.ObjectID) uint
 
-	// Only update the generation numbers map if the connection has retrieved its generation number. Otherwise, we'd
-	// decrement the count for the generation even though it had never been incremented.
-	if c.hasGenerationNumber() {
-		c.descMu.RLock()
-		serviceID := c.desc.ServiceID
-		c.descMu.RUnlock()
-		p.generation.removeConnection(serviceID)
+	// Only update the generation numbers map if the connection has retrieved its generation number.
+	// Otherwise, we'd decrement the count for the generation even though it had never been
+	// incremented.
+	if conn.hasGenerationNumber() {
+		p.generation.removeConnection(conn.desc.ServiceID)
 	}
 
-	if publishEvent && p.monitor != nil {
-		c.pool.monitor.Event(&event.PoolEvent{
+	if p.monitor != nil {
+		p.monitor.Event(&event.PoolEvent{
 			Type:         event.ConnectionClosed,
-			Address:      c.pool.address.String(),
-			ConnectionID: c.poolID,
+			Address:      p.address.String(),
+			ConnectionID: conn.poolID,
 			Reason:       reason,
 		})
 	}
+
 	return nil
 }
 
-// put returns a connection to this pool. If the pool is connected, the connection is not
-// stale, and there is space in the cache, the connection is returned to the cache. This
-// assumes that the connection has already been counted in p.conns.totalSize.
-func (p *pool) put(c *connection) error {
-	defer p.sem.Release(1)
+// checkIn returns an idle connection to the pool. If the connection is perished or the pool is
+// disconnected, it is removed from the connection pool and closed.
+func (p *pool) checkIn(conn *connection) error {
+	if conn == nil {
+		return nil
+	}
+	if conn.pool != p {
+		return ErrWrongPool
+	}
+
 	if p.monitor != nil {
-		var cid uint64
-		var addr string
-		if c != nil {
-			cid = c.poolID
-			addr = c.addr.String()
-		}
 		p.monitor.Event(&event.PoolEvent{
 			Type:         event.ConnectionReturned,
-			ConnectionID: cid,
-			Address:      addr,
+			ConnectionID: conn.poolID,
+			Address:      conn.addr.String(),
 		})
 	}
 
-	if c == nil {
+	return p.checkInNoEvent(conn)
+}
+
+// checkInNoEvent returns a connection to the pool. It behaves identically to checkIn except it does
+// not publish events. It is only intended for use by pool-internal functions.
+func (p *pool) checkInNoEvent(conn *connection) error {
+	if conn == nil {
 		return nil
 	}
-
-	if c.pool != p {
+	if conn.pool != p {
 		return ErrWrongPool
 	}
 
-	_ = p.conns.Put(c)
+	if reason, perished := connectionPerished(conn); perished {
+		_ = p.removeConnection(conn, reason)
+		go func() {
+			_ = p.closeConnection(conn)
+		}()
+		return nil
+	}
+
+	p.idleMu.Lock()
+	defer p.idleMu.Unlock()
 
+	for {
+		w := p.idleConnWait.popFront()
+		if w == nil {
+			break
+		}
+		if w.tryDeliver(conn, nil) {
+			return nil
+		}
+	}
+
+	for _, idle := range p.idleConns {
+		if idle == conn {
+			return fmt.Errorf("duplicate idle conn %p in idle connections stack", conn)
+		}
+	}
+
+	p.idleConns = append(p.idleConns, conn)
 	return nil
 }
 
```

Comment:
I chose to maintain the behavior of the existing `pool.clear()` implementation here because it's not clear that changing it to align with the CMAP spec would help resolve connection creation problems. The [GODRIVER-1799](https://jira.mongodb.org/browse/GODRIVER-1799) epic seems to make sense to align `pool` behavior with the CMAP spec, although there's no ticket that directly addresses making `pool.clear()` clear the WaitQueue.

I will create a ticket that describes updating `pool.clear()` behave like the CMAP spec describes.

Reply:
Created [GODRIVER-2165](https://jira.mongodb.org/browse/GODRIVER-2165) to track updating `pool.clear()` behavior.